### DISCUSSION TITLE
idea about mapping macros and more

### DISCUSSION THREAD

**Discussion by TheFeelipe** [stephenberry/glaze#908] [Ideas]
Is it possible to create macros to facilitate struct mapping, something similar to NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT from the nlohmann library?

current manual mapping is very verbose and tiring.

Something that could also help would be the possibility of using glz::read<T> with C arrays for legacy projects, I saw that in auto mapping it is possible to create a C char array, this is very good, I could bring this to the others also, even if it was initially manual mapping

**Comment by stephenberry** [comment]
Glaze had these macros ([see old file](https://github.com/stephenberry/glaze/blob/v2.0.0/include/glaze/core/macros.hpp)), but intentionally deprecated them not long ago.

The nlohmann examples of using this macro [NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT](https://json.nlohmann.me/api/macros/nlohmann_define_type_non_intrusive/#examples) can be purely reflection with Glaze and don't require any typing at all.

In cases where `glz::meta` or the local `struct glaze` are required, the non-macro approach is extremely powerful and not limited like the macros. We've found that the macros often had to be refactored into the more flexible meta approach when adding more complex fields to classes, which was annoying and added more complexity in the long run.

If you could provide an example of what you consider to be verbose, there may be ways to simplify the code. Or, you could rename the old glaze macros and use them yourself if you wanted.

Some reasons to avoid the macros:
- They will become more obsolete as we gain more reflection and glaze reflection improves
- We found that macros increased compile times for large objects
- They don't allowing selectively renaming key names
- They don't allow glaze wrappers

Another major reason to not use macros is that we want to support C++20 modules, which disallows exporting macros.

On your other point of supporting glz::read with C arrays, this is very desirable and I've opened an issue here #909. I will note that you can use lambda wrappers with `std::span` to currently support C arrays. But, that is certainly verbose.

```c++
struct my_struct
{
   uint32_t array[4]{};

   struct glaze {
      static constexpr auto value = glz::object("array", [](auto& s){ return std::span{ s.array, 4 }; });
   };
};
```

**Reply by stephenberry** [reply]
@TheFeelipe, I just merged support for C style arrays in #910
